wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationParticipantsService.swift

```diff
@@ -126,7 +117,45 @@ public class ConversationParticipantsService: ConversationParticipantsServiceInt
             try await proteusParticipantsService.addParticipants(users, to: conversation)
 
             // For mixed protocol we only try once and don't handle errors
-            try? await mlsParticipantsService.addParticipants(users, to: conversation)
+            try? await addMLSParticipants(users, to: conversation, showErrors: false)
+        }
+    }
+
+    private func addMLSParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation,
+        showErrors: Bool
+    ) async throws {
+
+        guard let mlsParticipantsService else {
+            throw ConversationParticipantsError.missingMLSParticipantsService
+        }
+
+        do {
+            try await mlsParticipantsService.addParticipants(users, to: conversation)
+        } catch MLSConversationParticipantsError.failedToClaimKeyPackages(users: let failedUsers) {
+
+            guard failedUsers.isNonEmpty else {
+                return
+            }
+
+            let users = Set(users)
+
+            if failedUsers != users {
+
+                // Operation was aborted because some users didn't have key packages
+                // We filter them out and retry once
+
+                try await internalAddParticipants(
+                    Array(users.subtracting(failedUsers)),
+                    to: conversation
+                )
+            }
+
+            if showErrors {
```

Comment:
I wouldn't catch the error here but rather bubble up handle it at the same level as `handleFederationError` and re-use the logic for inserting system message.

I understand that isn't in the scope of this PR, but how about letting the error bubble up here and catch and swallow it when adding participants in mixed mode and leave the error handling for WPB-2228

Reply:
like that https://github.com/wireapp/wire-ios/pull/699/commits/327a79b95064e073b2204031f8a1242b1d7704ed ?

Reply:
Yes